src/download.rs

```diff
@@ -0,0 +1,107 @@
+//! Represents a file to be downloaded.
+
+use crate::Error;
+use std::convert::TryFrom;
+use url::Url;
+
+/// Represents a file to be downloaded.
+#[derive(Debug, Clone)]
+pub struct Download {
+    /// URL of the file to download.
+    pub url: Url,
+    /// Filename used to save the file on disk.
+    pub filename: String,
+}
+
+impl Download {
+    /// Creates a new [`Download`].
+    ///
+    /// When using the [`Download::try_from`] method, the filename is
+    /// automatically extracted from the URL.
+    ///
+    /// ## Example
+    ///
+    /// The following calls are equivalent, minus some extra URL validations
+    /// performed by `try_from`:
+    ///
+    /// ```no_run
+    /// # use color_eyre::{eyre::Report, Result};
+    /// use trauma::download::Download;
+    /// use url::Url;
+    ///
+    /// # fn main() -> Result<(), Report> {
+    /// Download::try_from("https://example.com/file-0.1.2.zip")?;
+    /// Download::new(&Url::parse("https://example.com/file-0.1.2.zip")?, "file-0.1.2.zip");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new(url: &Url, filename: &str) -> Self {
+        Self {
+            url: url.clone(),
+            filename: String::from(filename),
+        }
+    }
+}
+
+impl TryFrom<&Url> for Download {
+    type Error = crate::Error;
+
+    fn try_from(value: &Url) -> Result<Self, Self::Error> {
+        let path_segments = value.path_segments();
+        if let Some(u) = path_segments {
+            if let Some(filename) = u.last() {
+                Ok(Download {
+                    url: value.clone(),
+                    filename: String::from(filename),
+                })
+            } else {
+                Err(Error::InvalidUrl(format!(
+                    "the url \"{}\" does not contain a filename",
+                    value
+                )))
+            }
+        } else {
+            Err(Error::InvalidUrl(format!(
+                "the url \"{}\" does not contain a valid path",
+                value
+            )))
+        }
+    }
+}
+
+impl TryFrom<&str> for Download {
+    type Error = crate::Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        let u = match Url::parse(value) {
```

Comment:
[`.map_err`](https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.map_err) can be used here

Reply:
Nice! Much clearer indeed! 

I made it work with the combinators:

```rust
Url::parse(value)
            .map_err(|e| {
                Error::InvalidUrl(format!("the url \"{}\" cannot be parsed: {}", value, e))
            })
            .and_then(|u| Download::try_from(&u))
```